### PR TITLE
WIP: Implemented Skia DPI scaling.

### PR DESCRIPTION
**Pull request by danwalmsley**

### CODE REVIEW

File: src/Skia/Avalonia.Skia/DrawingContextImpl.cs [hunk 1]
```diff
@@ -11,13 +11,19 @@ namespace Avalonia.Skia
...
 
-        public DrawingContextImpl(SKCanvas canvas)
+        public DrawingContextImpl(SKCanvas canvas, double scale = 1.0)
         {
             Canvas = canvas;
             Canvas.Clear();
+
+            _dpiScaleMatrix = Matrix.CreateScale(scale, scale);
+
+            _currentTransform =  Matrix.Identity * _dpiScaleMatrix;
```
Comment (grokys) [inline src/Skia/Avalonia.Skia/DrawingContextImpl.cs]: You can think of `Matrix.Identity` as a 1 in a multiplication, so you don't need it here, as essentially you're saying `currentTranform = 1 * scale`. In other words ` Matrix.Identity * _dpiScaleMatrix == _dpiScaleMatrix`.

Comment (grokys) [inline src/Skia/Avalonia.Skia/DrawingContextImpl.cs]: However, you're not informing the `Canvas` of the matrix here...

File: src/Skia/Avalonia.Skia/DrawingContextImpl.cs [hunk 2]
```diff
@@ -355,7 +359,11 @@ public Matrix Transform
                     return;
 
                 _currentTransform = value;
-                Canvas.SetMatrix(value.ToSKMatrix());
+
+                // TODO this line is a workaround for bug in DrawingContext.cs
+                _currentTransform *= _dpiScaleMatrix;
```
Comment (grokys) [inline src/Skia/Avalonia.Skia/DrawingContextImpl.cs]: ...which is why I suspect you need this here. This is actually not doing the right thing: it's applying the scaling *after* the current transform, whereas it needs to be applied before.

**Review by grokys — CHANGES_REQUESTED** [review]

File: src/Skia/Avalonia.Skia.Desktop/RenderTarget.cs [hunk 3]
```diff
@@ -79,6 +79,31 @@ private void GetPlatformWindowSize(out int w, out int h)
 #endif
         }
 
+#if WIN32
```
Comment (kekekeks) [inline src/Skia/Avalonia.Skia.Desktop/RenderTarget.cs]: It's definitely not the job of rendering system to detect window DPI.

**Comment by kekekeks** [comment]
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.SceneGraph/Media/DrawingContext.cs
This class is specifically designed to allow arbitrary transformations to be applied without touching underlying implementation. Just use `PushPostTransform` or `PushPreTransform` before returning the context instead of applying hacks like you do in this PR

**Comment by danwalmsley** [comment]
@kekekeks 

" Just use PushPostTransform or PushPreTransform before returning the context instead of applying hacks like you do in this PR"

this didn't work.

 because:
https://github.com/AvaloniaUI/Avalonia/blob/85f07274d434b11d9aa722a491c2286fcf0791d7/src/Avalonia.SceneGraph/Rendering/RendererMixin.cs#L128
causes the scaled matrix to end up in RenderTransforms and it affects hittesting and adorner layers...
set here.... (and probably other places)
https://github.com/AvaloniaUI/Avalonia/blob/85f07274d434b11d9aa722a491c2286fcf0791d7/src/Avalonia.Controls/Primitives/AdornerLayer.cs#L62
any suggestions how I can resolve this?
I guess we really want to completely separate out DPI scaling

**Comment by kekekeks** [comment]
@danwalmsley It *can't* affect render transforms in controls from there.

**Pull request merged**